pkg/controller/machine/queue.go

```diff
@@ -0,0 +1,53 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machine
+
+import (
+	"time"
+
+	"github.com/golang/glog"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	"github.com/kubernetes-incubator/apiserver-builder/pkg/controller"
+	"sigs.k8s.io/cluster-api/pkg/controller/config"
+)
+
+func (c *MachineController) RunAsync(stopCh <-chan struct{}) {
+	for _, q := range c.Informers.WorkerQueues {
+		c.StartWorkerQueue(q, stopCh)
+	}
+	controller.GetDefaults(c.controller).Run(stopCh)
+}
+
+// StartWorkerQueue schedules a routine to continuously process Queue messages
+// until shutdown is closed
+func (c *MachineController) StartWorkerQueue(q *controller.QueueWorker, shutdown <-chan struct{}) {
+	glog.Infof("Start %s Queue", q.Name)
+
+	for i := 0; i < config.ControllerConfig.ThreadCount; i++ {
+		// Every second, process all messages in the Queue until it is time to shutdown
+		go wait.Until(func() { q.ProcessAllMessages() }, time.Second, shutdown)
```

Comment:
This is significantly simpler than what you had at first. Could you briefly explain why what you had before is not necessary and this is sufficient? Does the queue make sure a machine can't be queued more than once? It looks like maybe, but I'm not sure.

Also, do you need the anonymous function as the first argument to `wait.Until`? Or can you pass `q.ProcessAllMessages` directly?

Reply:
The queue package in client-go actually provided native support by allowing only one reconciliation active per resource. Checking out the Add() routine (https://github.com/kubernetes/client-go/blob/master/util/workqueue/queue.go), we don't really enqueue a new task if it's already in-progress. Instead, we marked the dirty bit. In Done() routine, we will requeue the task if it's marked as dirty. It's quite a clever and straightforward design.